### DISCUSSION TITLE
Some possible JS improvements

### DISCUSSION THREAD

**Discussion by irskep** [cu/silicon#15] [General]
**Replacing requirejs with native module imports**

I took a quick crack at removing the requirejs dependency and using [import maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps) to get CodeMirror. It didn't work. It seems possible to [import CodeMirror from esm.sh](https://discuss.codemirror.net/t/esm-compatible-codemirror-build-directly-importable-in-browser/5933/11), but requiring a third party CDN seems like a non-starter.

It's possible CodeMirror will eventually publish a natively-importable bundle. Until then, requirejs it is.

However, it probably makes sense to use import maps and native module imports for the internal JS files, rather than stacking up `<script>` tags.

**Single entry point**

If you have a single `main.js` file that is added to every page, it's easier to work on the JS without needing to inspect the HTML templates to see which scripts are imported. The JS functions all operate on specific elements on the page, so it shouldn't be dangerous to over-include stuff.

**Formatter**

Everybody has idiosyncratic formatting preferences. As the repo owner, you can enforce your own by adding a `.prettierrc`. I'd be happy to write one for you to fit your needs. Do you have any rules in mind or could we just apply the default settings and let it reformat a bit? (It looks like `.editorconfig` wants your JS to use 2-space indents but it uses 4-space, I assume 4 is the right answer?)

**Camel case**

I haven't seen a JS project that uses snake case in years. :-) Are you open to camel case just to reduce the likelihood that someone goes against the style out of sheer muscle memory?

**Comment by cu** [comment]
A couple months back, I started getting serious about learning JS properly. I'm still only maybe 50% of the way through learning all that I want to. In the meantime, I'm happy to have help getting things steered in the right direction by someone with more experience. :)

Re CodeMirror, I believe I ran into the same thing. 

I'm totally open to the rest of your suggestions.

> 4 spaces and camel case

These were just me imposing my prior programming preferences on the JS code as an act of rebellion. But I've grown up a little and I'm okay with standardizing on whatever the prevalent JS practices of 2024 are. (Minus adding a bundler or whatever, if it can be avoided.)

**Reply by irskep** [reply]
I'm also a fan of 2 spaces! We can configure it that way with prettierrc and then everybody who uses vscode will do the right thing automatically.

I would encourage you to consider letting [esbuild](https://esbuild.github.io) into your heart, just until you can import all your dependencies via the browser. It uses the same syntax as the web standard, uses a single binary, and maintains a tight scope of responsibility instead of trying to be the kitchen sink. As it is, you're saying "I don't want the workflow complexity of a bundler," but the tradeoff is that the requirejs workflow is even more idiosyncratic without really giving you anything in return. It also cuts you off from being able to search the web for solutions to your problems, since you need to always take this unusual path to accomplishing common tasks.

So my suggestion is you draw the line at "ES6 only, no JSX or TypeScript" to allow you to use CodeMirror more straightforwardly, and potentially pull in other dependencies that are directly beneficial. This approach relies only on open standards and not special bundler features.

Consider all this as advice but not argument. I totally get the motivation to avoid unnecessary tools.

**Reply by irskep** [reply]
One more note about bundlers: when I use esbuild in personal projects, I tend to just commit the output to git because it's relatively readable. You don't need to make it a dependency for running the project, just developing the JS.

**Reply by irskep** [reply]
...and another thought: bundling your JS would remove the need for new contributors to do the `npm install` dance in order to use the CodeMirror editor. You'd only need to install the node toolchain for working on the JS.

Maybe you could maintain a 'barebones' set of build-less JS files, and an 'augment' set of bundled files that can use NPM packages?

**Reply by cu** [reply]
I think for now I will commit to trying out esbuild once I get a little further along in learning JS. If it's true that it can be used without being a "real" dependency, that's a strong positive. I appreciate the advice.

**Comment by irskep** [comment]
Lack of a bundler turns out to be a blocker for using CodeMirror themes, which is necessary for fully correct dark mode support.

What you'd normally do is import `@codemirror/theme-one-dark` and use it in the configuration. But that library doesn't seem to play well with requirejs. I can inline the code, but that's introducing new complexity. JS libraries don't generally expect to be used in this way; requirejs is not a browser standard.

Maybe this is possible and I couldn't figure it out, but that in itself is a signal.

**Comment by cu** [comment]
Just to let you know, this hasn't fallen off my radar... I've been making slow but steady progress getting more up to speed on modern JS and development practices while working on another project. Once that one is more or less functional, I plan to circle back to Silicon to refactor a few things based on your suggestions plus some features that I've been wanting to add. I don't know how long that will take but I just wanted to note that I wouldn't be at all offended if you wanted to fork Silicon in order to make progress on your own improvements without my limited availability holding you up.
